cirq/sim/clifford/stabilizer_sampler.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List
+
+import numpy as np
+
+import cirq
+from cirq import circuits, protocols, value
+from cirq.sim.clifford.act_on_clifford_tableau_args import \
+    ActOnCliffordTableauArgs
+from cirq.sim.clifford.clifford_tableau import CliffordTableau
+from cirq.work import sampler
+
+
+class StabilizerSampler(sampler.Sampler):
+    """An efficient sampler for stabilizer circuits."""
+
+    def __init__(self, *, seed: 'cirq.RANDOM_STATE_OR_SEED_LIKE' = None):
+        """
+        Args:
+            seed: The random seed or generator to use when sampling.
+        """
+        self.init = True
+        self._prng = value.parse_random_state(seed)
+
+    def run_sweep(
+            self,
+            program: 'cirq.Circuit',
+            params: 'cirq.Sweepable',
+            repetitions: int = 1,
+    ) -> List['cirq.TrialResult']:
+        trial_results: List[cirq.TrialResult] = []
+        for param_resolver in cirq.to_resolvers(params):
+            resolved_circuit = cirq.resolve_parameters(program, param_resolver)
+            measurements = self._run(
+                resolved_circuit,
+                repetitions=repetitions,
+            )
+            trial_results.append(
+                cirq.TrialResult(params=param_resolver,
+                                 measurements=measurements))
+        return trial_results
+
+    def _run(self, circuit: circuits.Circuit,
+             repetitions: int) -> Dict[str, np.ndarray]:
+
+        measurements: Dict[str, List[int]] = {
+            key: [] for key in protocols.measurement_keys(circuit)
+        }
+        axes_map = {q: i for i, q in enumerate(circuit.all_qubits())}
+
+        for _ in range(repetitions):
```

Comment:
Question: Why not propagate the stabilizer once and sample from the resulting stabilizer state at the end? I suppose it's simply deferred to a follow-up?

Reply:
Measurements can return random results which play out in different ways. In principle we could check if all measurements are terminal, or translate the circuit into a graph dataflow form that was faster to sampler from (assuming all classically controlled operations were Paulis). I wasn't planning to do these optimizations.

Reply:
The optimization would mostly target circuits with terminal measurements. We could also support measurements commuting with stabilizer, but that may be excessive and not useful. Finally, in some applications you want probability of a prescribed (Pauli) measurement result and in this case there is a unique update to the stabilizer generators consistent with the result.